### DISCUSSION TITLE
What algorithm is used to draw smooth curves in real time?

### DISCUSSION THREAD

**Discussion by sergey900553** [steveruizok/perfect-freehand#61] [Q&A]
Perfect freehand creates smooth curves in real time. In other applications, this requires some kind of post-processing.

It looks like some kind of spline, like a Catmull-Rom spline. But this app shows too good results. What algorithm are you using for this?

### Perfect freehand  smoothes in real time.
![bandicam 2022-10-28 09-40-59-325](https://user-images.githubusercontent.com/42143402/198520559-b9f81f4a-160d-47f9-b7c2-932fc03d7256.gif)

### Another application uses post-processing
![bandicam 2022-10-25 16-53-09-288 (1)](https://user-images.githubusercontent.com/42143402/198520830-2dcdc60f-024f-4334-b34a-d5c48fb87866.gif)

**Comment by colbyn** [comment]
@sergey900553  I don't think perfect-freehand is using any such formal algorithm, if you look at the implementation everything looks relatively ad-hoc (although there is a remark that comes to mind about smoothing points along a more "perfect curve").

**Reply by GreatTA1998** [reply]
Which specific files contain the implementation detail for perfect-freehand's continuous smoothing?

**Comment by colbyn** [comment]
Also for context if it helps, the perfect-freehand algorithm basically just gives you offset points, and you don't even need to render stroke outline point as as curves, if you have a high enough sampling frequency (and I suppose if the stroke diameter isn't too big) lines will be indistinguishable from curves.
